File: LeetCode/1334.Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance/Mario/1334.Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance.java
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] matrix = new int[n][n];
        for (int[] row : matrix) {
            Arrays.fill(row, 10001);
        }
        for (int i = 0; i < n; i++) {
            matrix[i][i] = 0;
        }
        for (int[] edge : edges) {
            int v1 = edge[0], v2 = edge[1], w = edge[2];
            matrix[v1][v2] = w;
            matrix[v2][v1] = w;
        }

        // Floyd algorithm
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    matrix[i][j] = Math.min(matrix[i][j], matrix[i][k] + matrix[k][j]);
                }
            }
        }

        int city = 0, minNei = n;
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] <= distanceThreshold) {
                    count += 1;
                }
                if (count > minNei)
                    break;
            }
            if (count <= minNei) {
                city = i;
                minNei = count;
            }
        }

        return city;
    }
}

File: LeetCode/1334.Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance/1334.Find_the_City_With_the_Smallest_Number_of_Neighbors_at_a_Threshold_Distance.py
#!/usr/bin/env python3

from typing import List
from math import inf


class Solution:
    def findTheCity(
        self, n: int, edges: List[List[int]], distanceThreshold: int
    ) -> int:
        # n cities, [0, n-1]
        # edges bidirectional
        # at most threshold

        dis = [[inf] * n for _ in range(n)]

        for i in range(n):
            dis[i][i] = 0

        for u, v, w in edges:
            dis[u][v] = w
            dis[v][u] = w

        # main Floyd Warshall
        for k in range(n):
            for i in range(n):
                dik = dis[i][k]
                if dik == inf:
                    continue
                for j in range(n):
                    if dis[k][j] != inf:
                        dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j])

        min_count = inf
        res = -1
        for i in range(n):
            cnt = 0
            for j in range(n):
                if i != j and dis[i][j] <= distanceThreshold:
                    cnt += 1

            if cnt <= min_count:
                min_count = cnt
                res = i

        return res
